File: TOP150/Minimum Path Sum.py

class Solution:

    def minPathSum(self, grid: List[List[int]]) -> int:
        
        m = len(grid) 
        n = len(grid[0]) 

        # def : dp[i][j] : minimum cost to arrive grid[i][j] 
        dp = [ [ None for i in range(n)] for j in range(m) ] 

        # base-case : dp[0][0] = grid[0][0] 
        dp[0][0] = grid[0][0] 

        for i in range(m) : 
            for j in range(n):  

                if i == 0 and j == 0 : pass 
                elif i==0 :  
                    dp[i][j] = dp [i][j-1]  + grid[i][j] 
                elif j == 0 : 
                    dp[i][j] = dp[i-1][j] + grid[i][j ] 
                else : 
                    dp[i][j] = grid[i][j] + min( dp[i-1][j] , dp[i][j-1]  )

        return dp[m-1][n-1] 


# State compression
class Solution:

    def minPathSum(self, grid: List[List[int]]) -> int:
                 
        m = len(grid) 
        n = len(grid[0]) 

        # first row :  
        prev_row  = [None for i in range(n)]
        temp = 0 
        for i in range(n): 
            temp += grid[0][i]
            prev_row[i] =  temp  

        for i in range(1 , m): 

            new_row = [None for i in range(n)]  

            for j in range(n): 

                if j == 0 : 
                    new_row[j] = prev_row[j] + grid[i][j] 
                else : 
                    new_row[j] = grid[i][j] + min(new_row[j-1] , prev_row[j]) 

            prev_row = new_row 

        return prev_row[-1] 
                    

File: RUSH/Minimum Path Sum.js
/**
 * @param {number[][]} grid
 * @return {number}
 */

// create 2D dp-table represent the minimum cost to arrive (i,j) ; 
var minPathSum = function(grid) {
    
    let dp = Array.from( new Array( grid.length ) , () => new Array(grid[0].length) ) ; 
    dp[0][0] = grid[0][0] ; 

    for (let i = 0 ; i < grid.length ; i++){

        for (let j = 0 ; j < grid[0].length ; j++){

            if (i===0 && j===0){
                dp[i][j] = grid[i][j] ; 
            }
            else if(i===0){ 
                dp[0][j] = grid[i][j] + dp[0][j-1] ; 
            }
            else if (j===0){
                dp[i][j]  = grid[i][j] + dp[i-1][j] ; 
            }
            else{
                dp[i][j] = grid[i][j] + Math.min(dp[i-1][j] , dp[i][j-1]) ; 
            }
        }
    }
    return dp[grid.length-1][grid[0].length-1]; 
};


// create 2D dp-table represent the minimum cost to arrive (i,j) ; 
// extreme state compress solution 
var minPathSum = function(grid) {

    let prevRow = new Array(grid.length) ; 

    prevRow[0] = grid[0][0] ; 
    // calculate first row 
    for (let j = 1 ; j < grid[0].length ; j++){ prevRow[j] = grid[0][j] + prevRow[j-1]  }


    for (let i = 1 ; i < grid.length ; i++){

        let leftElement = null ; 

        for (let j = 0 ; j < grid[0].length ; j++){

            if (leftElement === null){
                prevRow[j] = grid[i][j] + prevRow[j] ;                 
                leftElement = prevRow[j] ; 
            }
            else{
                prevRow[j] = grid[i][j] + Math.min( prevRow[j-1] , prevRow[j] )
                leftElement = prevRow[j] ; 
            }
        }
    }

    return prevRow[grid[0].length-1]

}